### ISSUE TITLE
Surprising behavior with `ExceptionInLinkedThread` from `async`

### ISSUE THREAD

**Issue by bitc** [fpco/safe-exceptions#38]
Is [ExceptionInLinkedThread](https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html#t:ExceptionInLinkedThread) a sync or async exception?

The answer according to this library is that it is async. Here is an example program:

```haskell
{-# LANGUAGE ScopedTypeVariables #-}

module Main where

import Control.Concurrent.Async
import Control.Concurrent (threadDelay)
import Control.Exception (SomeException)
import Control.Exception.Safe (try, throwString)

main = do
  result1 <- try doSomething1
  case result1 of
    Left (ex :: SomeException) -> putStrLn "1 Error"
    Right () -> putStrLn "1 Success"

  result2 <- try doSomething2
  case result2 of
    Left (ex :: SomeException) -> putStrLn "2 Error"
    Right () -> putStrLn "2 Success"

doSomething1 :: IO ()
doSomething1 = do
  putStrLn "1 Doing something..."
  throwString "Some Error"

doSomething2 :: IO ()
doSomething2 = do
  putStrLn "2 Doing something..."
  withAsync (threadDelay (1 * 1000 * 1000) >> throwString "Some Error") $ \a -> do
    link a
    threadDelay (2 * 1000 * 1000)
```

The output is:

```
1 Doing something...                                                                                                                                                                          
1 Error                                                                                                                                                                                       
2 Doing something...                                                                                                                                                                          
*** Exception: ExceptionInLinkedThread (ThreadId 168) Control.Exception.Safe.throwString called with:                                                                                         
                                                                                                                                                                                              
Some Error                                                                                                                                                                                    
Called from:                                                                                                                                                                                  
  throwString (Main.hs:29:47 in main:Hello)                                                                                                                                                  
```

The caller of `doSomething1`/`doSomthing2` wants to catch all (sync) exceptions that may happen inside those functions.

`doSomething1` is handled well, and the error is caught.

But `doSomething2`, despite being written in the recommended safe style according to the `async` documentation, allows the `ExceptionInLinkedThread` to slip past the "try" handler. This is not what we want, because the original caller shouldn't/doesn't care that "Some Error" exception was internally thrown on some linked thread.

One fix is for the author of `doSomething2` to wrap the entire function implementation with an exception handler that catches `ExceptionInLinkedThread` and then throws some other (sync) exception instead.

I tend to agree that `ExceptionInLinkedThread` should be considered by this library as an async exception, but I still think it can lead to surprising behavior, and require extra boilerplate which one might not expect to be needed. That is the reason I have opened this issue, to share my findings with others, but please feel free to promptly close this issue if it is concluded that nothing should be done.

Thank you

**Comment by snoyberg** [comment]
I think the behavior is exactly correct. The exception is in fact being throw to the current thread by another thread of execution, and therefore should not be recoverable. What's the difference between the `link` usage here versus a more general `throwTo` call?

**Comment by bitc** [comment]
> The exception is in fact being throw to the current thread by another thread of execution [...]

This I agree with.

> [...] and therefore should not be recoverable

This is debatable.

The example code I gave is similar to code that uses [race](https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html#v:race) or [concurrently](https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html#v:mapConcurrently). With these functions we also have an exception that originally happens on another thread being re-thrown on the main thread, and we DO want it recoverable.

Perhaps an example will be able to best illustrate my thoughts:

```haskell
module Test where

import Control.Concurrent.Async

downloadFile :: String -> IO String
downloadFile = error "TODO..."

main1 :: IO ()
main1 = do
  (file1, file2) <- concurrently (downloadFile "file1") (downloadFile "file2")
  print (file1, file2)

main2 :: IO ()
main2 = do
  (file1, file2) <- withAsync (downloadFile "file1") $ \a1 -> do
    link a1
    file2 <- downloadFile "file2"
    file1 <- wait a1
    pure (file1, file2)
  print (file1, file2)
```

There are two functions "main1" and "main2" that both do the same thing: download 2 files concurrently (and throw an exception *immediately* if there is a problem downloading either of the files)

The first implementation "main1" is implemented using the high-level "concurrently" function, and results in a sync exception ("recoverable") being thrown.

The second implementation "main2" decides to use the lower-level building blocks of the async library in order to achieve the same result. I wrote "main2" with the goal of behaving as close as possible to "main1", and I cannot think of an obvious change/improvement that should be done to "main2". However the subtle difference is that if there is a problem with "file1" then we get an `ExceptionInLinkedThread` exception, with the crucial part that this is an async exception that will not be caught by the caller of "main2". (This is compared with "main1" where it will be a sync exception that will be caught).

Anyway, after looking at this example I now realize that `ExceptionInLinkedThread` must remain async. If it were changed to be a sync exception, then if I were to put a catch-all sync exception handler around `downloadFile "file2"` then it might end up catching an exception from "file1", which would feel very very wrong.

So my conclusion is that nothing should be done about any of this and I will close this issue. Users of async should be aware of the behavior of `link`, and if they want to write code in the style of "main2" and want all exceptions to be recoverable, should wrap their code in a `catch` that catches `ExceptionInLinkedThread` and re-throws the inner exception.
